File: src/lib/projection/types/income.ts
import type { IsoDate } from "./model";

export const INCOME_DATA_PUBLIC_PATH = "/data/income";
export const INCOME_DATA_API_PATH = "/api/income-data";

export const INCOME_DATA_FILE_NAMES = {
	incomeSources: "income-sources.csv",
	taxProfiles: "tax-profiles.csv",
} as const;

export interface IncomeSourceDefinition {
	id: string;
	label: string;
	effectiveFrom: IsoDate;
	effectiveTo: IsoDate | null;
	annualGrossIncome: number;
}

export interface IncomeTaxBracket {
	upTo: number | null;
	rate: number;
}

export interface IncomeTaxProfile {
	id: string;
	label: string;
	deduction: number;
	brackets: IncomeTaxBracket[];
	sourceUrl: string | null;
}

export interface IncomeDataSnapshot {
	incomeSources: IncomeSourceDefinition[];
	taxProfiles: IncomeTaxProfile[];
}

export const EMPTY_INCOME_DATA: IncomeDataSnapshot = {
	incomeSources: [],
	taxProfiles: [],
};

File: backend/internal/types/income.go
package types

// IncomeSourceDefinition is one effective-dated annual gross income row.
type IncomeSourceDefinition struct {
	ID                string   `json:"id"`
	Label             string   `json:"label"`
	EffectiveFrom     IsoDate  `json:"effectiveFrom"`
	EffectiveTo       *IsoDate `json:"effectiveTo"`
	AnnualGrossIncome float64  `json:"annualGrossIncome"`
}

type IncomeTaxBracket struct {
	UpTo *float64 `json:"upTo"` // nil = open-ended
	Rate float64  `json:"rate"`
}

type IncomeTaxProfile struct {
	ID        string             `json:"id"`
	Label     string             `json:"label"`
	Deduction float64            `json:"deduction"`
	Brackets  []IncomeTaxBracket `json:"brackets"`
	SourceURL *string            `json:"sourceUrl"`
}

type IncomeDataSnapshot struct {
	IncomeSources []IncomeSourceDefinition `json:"incomeSources"`
	TaxProfiles   []IncomeTaxProfile       `json:"taxProfiles"`
}

func EmptyIncomeData() *IncomeDataSnapshot {
	return &IncomeDataSnapshot{
		IncomeSources: []IncomeSourceDefinition{},
		TaxProfiles:   []IncomeTaxProfile{},
	}
}
